crates/bevy_pbr/src/render/pbr_functions.wgsl

```diff
@@ -0,0 +1,602 @@
+    }
+
+    // compute texture coordinates for shadow lookup, compensating for the Y-flip difference
+    // between the NDC and texture coordinates
+    let flip_correction = vec2<f32>(0.5, -0.5);
+    let light_local = offset_position_ndc.xy * flip_correction + vec2<f32>(0.5, 0.5);
+
+    let depth = offset_position_ndc.z;
+    // do the lookup, using HW PCF and comparison
+    // NOTE: Due to non-uniform control flow above, we must use the level variant of the texture
+    // sampler to avoid use of implicit derivatives causing possible undefined behavior.
+#ifdef NO_ARRAY_TEXTURES_SUPPORT
+    return textureSampleCompareLevel(directional_shadow_textures, directional_shadow_textures_sampler, light_local, depth);
+#else
+    return textureSampleCompareLevel(directional_shadow_textures, directional_shadow_textures_sampler, light_local, i32(light_id), depth);
+#endif
+}
+
+fn hsv2rgb(hue: f32, saturation: f32, value: f32) -> vec3<f32> {
+    let rgb = clamp(
+        abs(
+            ((hue * 6.0 + vec3<f32>(0.0, 4.0, 2.0)) % 6.0) - 3.0
+        ) - 1.0,
+        vec3<f32>(0.0),
+        vec3<f32>(1.0)
+    );
+
+	return value * mix( vec3<f32>(1.0), rgb, vec3<f32>(saturation));
+}
+
+fn random1D(s: f32) -> f32 {
+    return fract(sin(s * 12.9898) * 43758.5453123);
+}
+
+// NOTE: This ensures that the world_normal is normalized and if
+// vertex tangents and normal maps then normal mapping may be applied.
+fn prepare_normal(
+    world_normal: vec3<f32>,
+#ifdef VERTEX_TANGENTS
+#ifdef STANDARDMATERIAL_NORMAL_MAP
+    world_tangent: vec4<f32>,
+#endif
+#endif
+    uv: vec2<f32>,
+    is_front: bool,
+) -> vec3<f32> {
+    var N: vec3<f32> = normalize(world_normal);
+
+#ifdef VERTEX_TANGENTS
+#ifdef STANDARDMATERIAL_NORMAL_MAP
+    var T: vec3<f32> = normalize(world_tangent.xyz - N * dot(world_tangent.xyz, N));
+    var B: vec3<f32> = cross(N, T) * world_tangent.w;
+#endif
+#endif
+
+    if ((material.flags & STANDARD_MATERIAL_FLAGS_DOUBLE_SIDED_BIT) != 0u) {
+        if (!is_front) {
+            N = -N;
+#ifdef VERTEX_TANGENTS
+#ifdef STANDARDMATERIAL_NORMAL_MAP
+            T = -T;
+            B = -B;
+#endif
+#endif
+        }
+    }
+
+#ifdef VERTEX_TANGENTS
+#ifdef STANDARDMATERIAL_NORMAL_MAP
+    let TBN = mat3x3<f32>(T, B, N);
+    N = TBN * normalize(textureSample(normal_map_texture, normal_map_sampler, uv).rgb * 2.0 - 1.0);
+#endif
+#endif
+
+    return N;
+}
+
+// NOTE: Correctly calculates the view vector depending on whether
+// the projection is orthographic or perspective.
+fn calculate_view(
+    world_position: vec4<f32>,
+) -> vec3<f32> {
+    var V: vec3<f32>;
+    let is_orthographic = view.projection[3].w == 1.0;
+    if (is_orthographic) {
+        V = normalize(vec3<f32>(view.view_proj[0].z, view.view_proj[1].z, view.view_proj[2].z));
+    } else {
+        // Only valid for a perspective projection
+        V = normalize(view.world_position.xyz - world_position.xyz);
+    }
+    return V;
+}
+
+struct PbrMaterial {
```

Comment:
The PbrMaterial vs StandardMaterial disconnect feels a bit off to me. We tend to use those words interchangeably and idk if introducing another "material" type to the mix is the right api. I get why StandardMaterial doesn't have an occlusion field (occlusion isn't really useful as a global setting).

Maybe we should just have a single PbrInput struct that include `material` and `occlusion` fields?

Reply:
So PbrInput with a StandardMaterial, occlusion, and the position/normal stuff? I'm fine with that.

Reply:
Yup!